rpc, test: Fix JSON parsing errors in unloadwallet and getdescriptoractivity RPCs

pablomartin4btc opened the pull request:
Currently, `unloadwallet` RPC call fails with a JSON parsing error when no `wallet_name` argument is provided. This behavior is misleading because the error originates from a low-level JSON type mismatch, rather than clearly indicating that the wallet name or RPC endpoint (`-rpcwallet=...`) is missing. Also, found out that the [issue](https://github.com/bitcoin/bitcoin/pull/13111#issuecomment-398831543) was noticed during its implementation but never addressed.

In addition, I've verified all RPC commands calls finding that `getdescriptoractivity` had the same problem, but related to the array input types (blockhashes & descriptors), so I've corrected that RPC as well. For consistency I've added the missing logging info for each test case in `test/functional/rpc_getdescriptoractivity.py` in preparation for the new test.

**_-Before_**
```
./build/bin/bitcoin-cli -regtest -datadir=/tmp/btc unloadwallet
error code: -3
error message:
JSON value of type number is not of expected type string
```
```
./build/bin/bitcoin-cli -regtest -datadir=/tmp/btc getdescriptoractivity
error code: -3
error message:
JSON value of type null is not of expected type array
```
```
./build/bin/bitcoin-cli -regtest -datadir=/tmp/btc getdescriptoractivity '[]'
error code: -3
error message:
JSON value of type null is not of expected type array
```
**_-After_**
```
./build/bin/bitcoin-cli -regtest -datadir=/tmp/btc unloadwallet
error code: -8
error message:
Either the RPC endpoint wallet or the wallet name parameter must be provided
```
```
./build/bin/bitcoin-cli -regtest -datadir=/tmp/btc getdescriptoractivity
error code: -1
error message:
getdescriptoractivity ["blockhash",...] [scanobjects,...] ( include_mempool )

Get spend and receive activity associated with a set of descriptors for a set of blocks. This command pairs well with the `relevant_blocks` output of `scanblocks()`.
This call may take several minutes. If you encounter timeouts, try specifying no RPC timeout (bitcoin-cli -rpcclienttimeout=0)

Arguments:
1. blockhashes                   (json array, required) The list of blockhashes to examine for activity. Order doesn't matter. Must be along main chain or an error is thrown.
                                 
     [
       "blockhash",              (string) A valid blockhash
       ...
     ]
2. scanobjects                   (json array, required) Array of scan objects. Every scan object is either a string descriptor or an object:
     [
       "descriptor",             (string) An output descriptor
       {                         (json object) An object with output descriptor and metadata
         "desc": "str",          (string, required) An output descriptor
         "range": n or [n,n],    (numeric or array, optional, default=1000) The range of HD chain indexes to explore (either end or [begin,end])
       },
       ...
     ]
3. include_mempool               (boolean, optional, default=true) Whether to include unconfirmed activity

...
```
```
./build/bin/bitcoin-cli -regtest -datadir=/tmp/btc getdescriptoractivity '[]'
error code: -1
error message:
getdescriptoractivity ["blockhash",...] [scanobjects,...] ( include_mempool )

...
```

pablomartin4btc committed:
test: Add missing logging info for each test

Add self.log.info(...) calls at the beginning of each test
in GetBlocksActivityTest.

This improves readability and provides debugging information
by logging the purpose of each test upon its correct
execution.

This is in preparation for the next commit, which adds a new test
with log info, and it would look inconsistent without this commit.

furszy reviewed (commented):
I think this worth a general RPC util function like (haven't tested it):

```c++
template <typename... Value>
bool AreParamsNullOrEmpty(const Value&... params) {
    return ((params.isNull() || params.empty()) && ...);
}
```

Which, for example should work fine for `getdescriptoractivity `:
```c++
if (AreParamsNullOrEmpty(request.params[0], request.params[1])) {
    throw JSONRPCError(RPC_INVALID_PARAMETER, "At least 1 blockhash or 1 descriptor must be specified.");
}
```

pablomartin4btc commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -462,6 +462,9 @@ static RPCHelpMan unloadwallet()
             throw JSONRPCError(RPC_INVALID_PARAMETER, "RPC endpoint wallet and wallet_name parameter specify different wallets");
         }
     } else {
+        if (request.params[0].isNull()) {
```
Also, RPC `migratewallet` calls the same function so I could refactor that instance too.

pablomartin4btc replied:
Done! Added you as co-author on that commit.

pablomartin4btc committed:
rpc, util: Add EnsureUniqueWalletName

Add a new function called EnsureUniqueWalletNamet that returns the
selected wallet name across the RPC request endpoint and wallet_name.

Supports the case where the wallet_name argument may be omitted—either
when using a wallet endpoint, or when not provided at all. In the latter
case, if no wallet endpoint is used, an error is raised.

Internally reuses the existing implementation to avoid redundant URL
decoding and logic duplication.

This is a preparatory change for upcoming refactoring of unloadwallet
and migratewallet, which will adopt EnsureUniqueWalletName for improved
clarity and consistency.

pablomartin4btc committed:
rpc, test: Fix error message in unloadwallet

The unloadwallet RPC previously failed with a low-level JSON parsing error
when called without any arguments (wallet_name).

Although this issue was first identified during review of the original unloadwallet
implementation in #13111, it was never addressed.

Raise RPC_INVALID_PARAMETER instead describing that either the RPC endpoint or wallet
name must be provided.

Adding a new functional test for this use case.

Refactor migratewallet to use the same logic as the wallet_name argument handling
is identical.

Co-authored-by:  maflcko <maflcko@users.noreply.github.com>

pablomartin4btc committed:
rpc, test: Fix error message in getdescriptoractivity

Mark blockhashes and scanobjects arguments as required, so the user receives
a clear help message when either is missing.

Added a new functional test for this use case.

Co-authored-by: stickies-v <stickies-v@users.noreply.github.com>

pablomartin4btc commented:
<ins>_Updates_</ins>:
- Addressed feedback from @maflcko (co-authored): extended `MaybeArg` to return a `std::optional<std::string>` for use in the `GetWalletNameFromJSONRPCRequest` overload. Also refactored the `migratewallet` handling of `wallet_name` to align with the existing behavior in `unloadwallet`.
- Incorporated @furszy's suggestion (co-authored): added a new template in RPC utils to check whether all provided parameters are null or empty, now used in the `getdescriptoractivity` RPC.

pablomartin4btc commented on src/rpc/util.cpp:
```diff
@@ -725,6 +725,7 @@ TMPL_INST(nullptr, const UniValue*, maybe_arg;);
 TMPL_INST(nullptr, std::optional<double>, maybe_arg ? std::optional{maybe_arg->get_real()} : std::nullopt;);
 TMPL_INST(nullptr, std::optional<bool>, maybe_arg ? std::optional{maybe_arg->get_bool()} : std::nullopt;);
 TMPL_INST(nullptr, const std::string*, maybe_arg ? &maybe_arg->get_str() : nullptr;);
+TMPL_INST(nullptr, std::optional<std::string>, maybe_arg ? std::optional{maybe_arg->get_str()} : std::nullopt;);
```
I do agree, just seemed more "idiomatic" to handle the wallet_name variable with the `has_value()`, but yeah, not very practical in the end. I've removed the `MaybeArg` extension commit making the `wallet_name` to hold a pointer in the `GetWalletNameFromJSONRPCRequest` overload.

pablomartin4btc replied:
I thought to use `<std::string_view>` too when I did the `TMPL_INST` but didn't go forward. We can keep it this way if this gets merged first or if it does yours I can incorporate it here yeah. I'll try to review your PR tmw. Thanks!

pablomartin4btc commented on src/rpc/util.h:
```diff
@@ -477,8 +477,8 @@ class RPCHelpMan
         auto i{GetParamIndex(key)};
         // Return optional argument (without default).
-        if constexpr (std::is_integral_v<R> || std::is_floating_point_v<R>) {
-            // Return numbers by value, wrapped in optional.
```
As in [previous](https://github.com/bitcoin/bitcoin/pull/32845#discussion_r2191837500) comment, already removed the commit containing also this change. I've analysed the [discussion](https://github.com/bitcoin/bitcoin/pull/28230#discussion_r1286815104) around the string copy in the original implementation of `MaybeArg` and even if we could go around it and "fix"/ adding it for the string specific case/ condition (~`return std::optional<std::string_view>{params[i].get_str_ref()}`?) perhaps makes it too complicated for this simple case that can be resolved much easier with the pointer as you said.

pablomartin4btc commented:
<ins>_Updates_</ins>
- Addressed @maflcko's feedback:
  - [dropped](https://github.com/bitcoin/bitcoin/pull/32845#discussion_r2191837500) `MaybeArg` extension commit, since a pointer already can hold the null state needed in the validations within the `GetWalletNameFromJSONRPCRequest` overload;
  - removed previously added `AreParamsNullOrEmpty` and replaced it with std::ranges::all_of in the `getdescriptoractivity` RPC fix instead, as [suggested](https://github.com/bitcoin/bitcoin/pull/32845#discussion_r2191846030).

furszy reviewed (commented):
utACK 529d2ce24570aae426ce19e87cb22f124dd09470

stickies-v commented on src/rpc/blockchain.cpp:
```diff
@@ -2669,6 +2669,11 @@ static RPCHelpMan getdescriptoractivity()
         },
         [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
 {
+    if (request.params[0].isNull() || request.params[1].isNull()) {
```
> Do we actually need to ensure that at least one such (valid) parameter is provided? If so, we should probably check that `blockindexes_sorted` and `scripts_to_watch` are not both empty. Otherwise, I think just making sure we only execute the `request.params[n].get_array().getValues()` statements if we've first checked that `!request.params[n].isNull()` should do the trick?

I don't think my comment was addressed, and e.g. this test still failing. According to the PR description and code changes it should pass since at least 1 blockhash or descriptor is passed.

<details>
<summary>git diff on 529d2ce245</summary>

```diff
diff --git a/test/functional/rpc_getdescriptoractivity.py b/test/functional/rpc_getdescriptoractivity.py
index 6dadfa10f6..5dac541e01 100755
--- a/test/functional/rpc_getdescriptoractivity.py
+++ b/test/functional/rpc_getdescriptoractivity.py
@@ -237,6 +237,7 @@ class GetBlocksActivityTest(BitcoinTestFramework):
         assert_raises_rpc_error(-8, "At least 1 blockhash or 1 descriptor must be specified.", node.getdescriptoractivity)
         self.log.info("Test that calling getdescriptoractivity with an empty array raises also RPC_INVALID_PARAMETER (-8)")
         assert_raises_rpc_error(-8, "At least 1 blockhash or 1 descriptor must be specified.", node.getdescriptoractivity, [])
+        node.getdescriptoractivity([self.generate(self.nodes[0], 1)[0]])
 
 
 if __name__ == '__main__':

```
</details>


The conceptual question raised in my earlier question was: do we just want to prevent `.get_array()` from throwing or do we actually need to ensure (valid or any) blockhashes or descriptors are passed. We'll need to answer this before we can decide on the best fix for this problem, I think. If the former, we might not need to raise any JSONRPCError. If the latter, we should probably inspect `blockindexes_sorted` and `scripts_to_watch`.

stickies-v replied:
> "descriptors" documentation could be improved, just quickly wrote something down. E.g. not sure if order matters here, or if there's anything else to look out for.

We should probably just leave the name as "scanobjects" (instead of my change to "descriptors"), to avoid introducing backwards incompatible API changes. The description should be change-able without issue.

stickies-v reviewed (commented):
I think the proposed fix for `getdescriptoractivity` is not robust, and the `GetWalletNameFromJSONRPCRequest()` overload confusing, as detailed in comments.

pablomartin4btc commented on src/wallet/rpc/util.h:
```diff
@@ -38,6 +38,7 @@ static const RPCResult RESULT_LAST_PROCESSED_BLOCK { RPCResult::Type::OBJ, "last
  */
 std::shared_ptr<CWallet> GetWalletForJSONRPCRequest(const JSONRPCRequest& request);
 bool GetWalletNameFromJSONRPCRequest(const JSONRPCRequest& request, std::string& wallet_name);
+std::string GetWalletNameFromJSONRPCRequest(const JSONRPCRequest& request, const std::string* wallet_name);
```
I could use a better naming for this function, not being an overload. My idea was to fix the issue without touching much and later refactor both `GetWalletNameFromJSONRPCRequest` and `GetWalletForJSONRPCRequest` on a separate PR. Currently `GetWalletForJSONRPCRequest`, which is used broadly, calls internally `GetWalletNameFromJSONRPCRequest` and wanted to keep them untouched for now.

pablomartin4btc replied:
Applied a very tiny tweak to it. Thanks!

pablomartin4btc commented:
<ins>_Updates_</ins>
- Addressed @stickies-v's feedback:
  - converted the `GetWalletNameFromJSONRPCRequest` overload to `EnsureUniqueWalletName` as [proposed](https://github.com/bitcoin/bitcoin/pull/32845#discussion_r2197447487) - no changes in logic;
  - [corrected](https://github.com/bitcoin/bitcoin/pull/32845#discussion_r2197391680) the `getdescriptoractivity` fix making both blockhashes and descriptors arrays as mandatory, returning the RPC Help for the command when any of them or both are missing. Updated the use cases in the PR description.

pablomartin4btc commented on src/rpc/blockchain.cpp:
```diff
@@ -2193,18 +2193,30 @@ static const auto scan_action_arg_desc = RPCArg{
+        }
+};
+
+static const auto output_descriptor = RPCArg{
+    "descriptor", RPCArg::Type::STR, RPCArg::Optional::OMITTED, "An output descriptor"
+};
+
+static const auto scanobjects_oneline_description = RPCArgOptions{
+    .oneline_description="[scanobjects,...]"
+};
```
> Would prefer keeping those 2 as-is?

Ok, I did it that way just for reuse purpose, to avoid having a typo on the strings and so on.

`obj_with_output_descriptor` comes from the definition: "An object with output descriptor and metadata", but I'm ok with `output_descriptor_obj`.

pablomartin4btc replied:
Done!

stickies-v commented on src/wallet/rpc/util.cpp:
```diff
@@ -29,6 +29,27 @@ bool GetAvoidReuseFlag(const CWallet& wallet, const UniValue& param) {
+        // wallet endpoint was used
+        if (wallet_name && *wallet_name != endpoint_wallet) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER,
+                "RPC endpoint wallet and wallet_name parameter specify different wallets");
```
Oh no I'm not talking about the helper function's parameter name, that's fine imo, I'm talking about the error message that's thrown and eventually returned to the user, referencing an RPC parameter that potentially doesn't exist. Again, no biggie, just a symptom that concerns aren't entirely separated.

BrandonOdiwuor reviewed (commented):
Tested ACK 01688e17534d3c9011cce92cff9f7935691bb80c

Was able to verify that this fixes the JSON parsing error on `unloadwallet` and `getdescriptoractivity` RPCs 

Tested on macOS 15.5

Branch: Master
<img width="870" height="203" alt="Screenshot 2025-07-16 at 11 41 41" src="https://github.com/user-attachments/assets/88508a93-3dfb-4676-9040-d8789677da1a" />

Commit: 01688e17534d3c9011cce92cff9f7935691bb80c
<img width="1511" height="887" alt="Screenshot 2025-07-15 at 14 36 44" src="https://github.com/user-attachments/assets/72cfcc8f-29fa-4fce-85a1-bbc1c48c6cdd" />
<img width="1512" height="922" alt="Screenshot 2025-07-15 at 14 38 06" src="https://github.com/user-attachments/assets/dbba14d8-859e-468f-b874-76d463978188" />

stickies-v commented on src/rpc/blockchain.cpp:
nit: I think the first paragraph of the 01688e17534d3c9011cce92cff9f7935691bb80c commit message can be improved to:

> Mark `blockhashes` and `scanobjects` arguments as required, so the user receives
> a clear help message when either is missing.

stickies-v commented on src/wallet/rpc/util.cpp:
```diff
@@ -29,6 +29,27 @@ bool GetAvoidReuseFlag(const CWallet& wallet, const UniValue& param) {
     return avoid_reuse;
 }
 
+std::string EnsureUniqueWalletName(const JSONRPCRequest& request, const std::string* wallet_name)
```
I wrote some unit tests for this function, feel free to add to this PR if you think it's helpful: https://github.com/stickies-v/bitcoin/commit/60f9dc8af66bbd4eda81049a10262527f83356a1

pablomartin4btc replied:
Thanks! Makes sense, I'll add them.

stickies-v reviewed (approved):
ACK 01688e17534d3c9011cce92cff9f7935691bb80c . Left a couple improvement suggestions that I think make sense to address here, but also aren't blocking. Happy to quickly re-review if you incorporate.

I think adding (very brief) release notes to indicate the slight behaviour change for `unloadwallet` and `getdescriptoractivity` could be useful for downstream projects.

pablomartin4btc committed:
rpc, test: Add EnsureUniqueWalletName tests

Co-authored-by: stickies-v <stickies-v@users.noreply.github.com>

pablomartin4btc committed:
doc: Add release notes for changes in RPCs

Adding notes for both `unloadwallet` and `getdescriptoractivity`.

pablomartin4btc commented on src/rpc/blockchain.cpp:
```diff
@@ -2632,10 +2644,13 @@ static RPCHelpMan getdescriptoractivity()
                 {"blockhash", RPCArg::Type::STR_HEX, RPCArg::Optional::OMITTED, "A valid blockhash"},
             }},
-            scan_objects_arg_desc,
+            RPCArg{"scanobjects", RPCArg::Type::ARR, RPCArg::Optional::NO, "Array of scan objects. Every scan object is either a string descriptor or an object:", {
```
Done!

pablomartin4btc commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -456,14 +456,7 @@ static RPCHelpMan unloadwallet()
-    } else {
-        wallet_name = request.params[0].get_str();
-    }
+    const std::string wallet_name{EnsureUniqueWalletName(request, self.MaybeArg<std::string>("wallet_name"))};
```
Done!

pablomartin4btc commented on test/functional/rpc_getdescriptoractivity.py:
```diff
@@ -221,6 +232,19 @@ def test_no_address(self, node, wallet):
+        assert_raises_rpc_error(-1, rpc_help_output, node.getdescriptoractivity)
+
+        self.log.info("Test that calling getdescriptoractivity with one required array missing raises also its RPC help")
+        assert_raises_rpc_error(-1, rpc_help_output, node.getdescriptoractivity, [])
```
Done!

pablomartin4btc commented:
<ins>_Updates_</ins>:
- Addressed [feedback](https://github.com/bitcoin/bitcoin/pull/32845#pullrequestreview-3024249604) from @stickies-v:
  - [Updated](https://github.com/bitcoin/bitcoin/pull/32845#discussion_r2204543796) `getdescriptoractivity` test for required params;
  - [Corrected](https://github.com/bitcoin/bitcoin/pull/32845#discussion_r2204560989) variables naming within RPC `getdescriptoractivity`;
  - [Updated](https://github.com/bitcoin/bitcoin/pull/32845#discussion_r2208503121) RPC invalid params error descriptions to be more generic and not depending on the RPC param naming representing the walle name;
  - [Updated](https://github.com/bitcoin/bitcoin/pull/32845#discussion_r2209965857) outdated docstring for `unloadwallet` RPC;
  - [Updated](https://github.com/bitcoin/bitcoin/pull/32845#discussion_r2209990945) `scanobjects` argument description within `getdescriptoractivity` RPC;
  - [Updated](https://github.com/bitcoin/bitcoin/pull/32845#discussion_r2210018576) `getdescriptoractivity` commit message;
  - [Incorporated](https://github.com/bitcoin/bitcoin/pull/32845#discussion_r2210134549) `EnsureUniqueWalletName` unit tests;
  - Added release notes for `unloadwallet` & `getdescriptoractivity` updates.

stickies-v commented on src/wallet/test/wallet_rpc_tests.cpp:
```diff
@@ -0,0 +1,41 @@
+#include <optional>
+#include <string>
+
+namespace wallet {
```
nit: what's the purpose of wrapping this in `namespace wallet`?

Relatedly, should probably make `TestWalletName` `static`.

pablomartin4btc replied:
Thanks for pointed that out. Forgot to remove `wallet::` in the `EnsureUniqueWalletName()` call. The reason I wrapped it in `namespace wallet` is that it helps to avoid verbose `wallet::` prefixes and reduces the chance of symbol collisions. It's also consistent with most of the other unit tests under `src/wallet/test`, which commonly wrap helper functions and test cases in the same namespace.

That said, I agree `TestWalletName()` should probably be marked `static` to indicate internal linkage — will fix that.

pablomartin4btc replied:
Done!

stickies-v replied:
> It's also consistent with most of the other unit tests under `src/wallet/test`

Fair enough, I wasn't aware.

> and reduces the chance of symbol collisions

I think all of this stuff is meant to have internal linkage, so if reducing symbol collisions is the goal (which seems sensible) an anonymous namespace might make more sense? Potentially with a `using namespace wallet`, even though at the moment we're only using one `wallet` symbol.

Anyway, not important, and following convention is usually a good default. Whatever you prefer.

stickies-v commented on src/rpc/blockchain.cpp:
```diff
@@ -2193,18 +2193,22 @@ static const auto scan_action_arg_desc = RPCArg{
+        output_descriptor_obj
     },
-    RPCArgOptions{.oneline_description="[scanobjects,...]"},
+    RPCArgOptions{.oneline_description="[scanobjects,...]"}
```
nit: can keep comma to avoid touching this line

pablomartin4btc replied:
Done!

stickies-v commented on doc/release-notes-32845.md:
```diff
@@ -0,0 +1,9 @@
+Updated RPCs
+------------
+
+- `unloadwallet` - Raise RPC_INVALID_PARAMETER describing that either the RPC endpoint
+or the wallet name parameter must be provided. Previously the RPC failed with a JSON
+parsing error.
```
```suggestion
- `unloadwallet` - Return RPC_INVALID_PARAMETER when both the RPC wallet endpoint
and wallet_name parameters are unspecified. Previously the RPC failed with a JSON
parsing error.
```

pablomartin4btc replied:
Done!

stickies-v reviewed (approved):
re-ACK af0ad72a0ed778c7e6f83b1a290705cefe23d78a

pablomartin4btc commented:
Addressed [feedback](https://github.com/bitcoin/bitcoin/pull/32845#pullrequestreview-3044915254) from @stickies-v.

stickies-v reviewed (approved):
re-ACK c5c1960f9350d6315cadbdc95fface5f85f25806

nit: clang-format doesn't agree with some of the spacing changes, if you force push again could be nice to run it on each commit to ensure consistency

furszy reviewed (commented):
ACK c5c1960f9350d6315cadbdc95fface5f85f25806
